README.md
# TemperatureSanitizer
TemperatureSanitizer detects when a minimum temperature is sustained for a minimum time. This program uses temperusb aka temper, and default settings are for killing bedbugs (set to 120mins though 90 is considered minimum, and 120 degrees fahrenheit though 118 is considered minimum).

In addition to the GPL 3.0 license, the following disclaimer applies: THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Requirements
* A TEMPerV1 compatible USB thermometer (can be ordered online)
* temperusb library for python (if your python does not have temperusb is not installed yet, the program will detect that problem and instruct you how to correct that). Installing temperusb requires the temperusb whl file or an internet connection in order to follow those instructions displayed by this program.

## Planned Features
* support TEMPered
  * Where `/dev/hidraw4` is the correct device (could be any number--usually last one listed via `ls /dev | grep hidraw`),
    run `sudo hid-query /dev/hidraw1 0x01 0x80 0x33 0x01 0x00 0x00 0x00 0x00`
    the response is 8 bytes such as:
    `80 80 0a fc  4e 20 00 00`
    where 0a fc is an integer (2825 in this case). Divide that by 100
    to get temperature.

  * an example of parsing the output is at <https://github.com/padelt/temper-python/issues/84#issuecomment-393930865>
  (possibly auto-install TEMPered such as with script below)
```bash
if [ ! `which hid-query` ]; then
  cd $HOME
  if [ ! -d Downloads ]; then mkdir Downloads; fi
  cd Downloads
  if [ ! -d TEMPered ]; then
    git clone https://github.com/edorfaus/TEMPered.git
    cd TEMPered
  else
    cd TEMPered
    git pull
  fi
  if [ ! -d build ]; then mkdir build; fi
  cd build
  cmake ..
  sudo make install
  if [ ! -d /usr/local/lib ]; then mkdir -p /usr/local/lib; fi
  sudo cp libtempered/libtempered.so.0 /usr/local/lib/
  sudo cp libtempered-util/libtempered-util.so.0 /usr/local/lib/
  sudo ln -s /usr/local/lib/libtempered.so.0 /usr/local/lib/libtempered.so
  sudo ln -s /usr/local/lib/libtempered-util.so.0 /usr/local/lib/libtempered-util.so
fi
```
  * find hid like:
```bash
  #LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib tempered
  # previous line doesn't work for some reason so:
  hid-query --enum
  # lists 2 for some reason:
  # /dev/hidraw0 : 413d:2107 interface 0 : (null) (null)
  # /dev/hidraw1 : 413d:2107 interface 1 : (null) (null)
```
  * make script named readTEMPer-driverless-withdate.sh (by jbeale1 from link above) like:
```bash
#!/bin/bash
OUTLINE=`sudo ./hid-query /dev/hidraw3 0x01 0x80 0x33 0x01 0x00 0x00 0x00 0x00|grep -A1 ^Response|tail -1`
OUTNUM=`echo $OUTLINE|sed -e 's/^[^0-9a-f]*[0-9a-f][0-9a-f] [0-9a-f][0-9a-f] \([0-9a-f][0-9a-f]\) \([0-9a-f][0-9a-f]\) .*$/0x\1\2/'`
HEX4=${OUTNUM:2:4}
DVAL=$(( 16#$HEX4 ))
bc <<< "scale=2; $DVAL/100"
```

  * use script above like
```bash
while [ true ]; do
  temp=`./readTEMPer-driverless-withdate.sh`
  echo $(date +"%F %T") " , " $temp
  sleep 14
done
```

## Known Issues
* Does not support TEMPer with USB ID 413d:2107 (try lsusb to see your ID while device is plugged in). This project should use https://github.com/edorfaus/TEMPered instead of temperusb (which has not responded to this issue on GitHub)

## Usage
* Make sure you place the temperature sensor in a "cold spot," such as (if you are using this program to monitor a bedbug oven) inside a book in the middle of your load.
* Change the following settings by editing the py file:
```python
desired_degrees = 120
desired_format = "fahrenheit"
interval_seconds = 60
desired_total_seconds = 120 * 60
```
    * and, if you want to chill instead of bake, you can edit the following settings also in the py file:
    ```python
    desired_comparison = ">="
    desired_collate_method = "average"
    ```
    * desired_comparison can be ">=", ">", "<=", or "<"
    * desired_collate_method can be minimum (or min), maximum (or max), or average (or avg) -- for example, for a chill (using < or <=), min or average would be appropriate comparisons to ensure that a temperature reading never was above the desired temperature; but for baking (using >= or >), min or average would be recommended, where min would ensure a temperature reading didn't dip below desired temperature.

* How the settings are used by the program:
    * desired_format can be "fahrenheit" or "celcius".
    * The interval may need to be at least a few seconds for accuracy, though an interval of 1 can be used if you don't mind having all that data, and your sensor is accurate enough for that type of use (trusting a single reading as part of your raw data).
    * The temperature is always checked every second. This is not configurable, but only the minimum and avarage are displayed from each interval, which is a group of these secondly readings.
    * The seconds will start counting when the minimum temperature of the interval is at least desired_degrees. If the total number of seconds of bake time is reached (when there was a consecutive set of spans where the minimum never was below desired_degrees), the program will end and show a summary (including a list of temperatures, each of which is the minimum temperature of each consecutive interval that had the desired temperature as its minimum). Otherwise, the program will continue collecting data indefinitely.
* Run:
```bash
# cd to the directory where you saved the py file, then:
sudo python TemperatureSanitizer
# Or, run without sudo if you have changed the permissions of your USB device.
# NOTE: sometimes that setting is forgotten when the device is unplugged and
# reinserted, even if same port is used.
```


## Changes
* (2017-04-09) Display remaining time while temperature is >= minimum
* (2017-04-09) Display system time when finished
* (2017-04-08) Renamed variables for clarity, and added a region called User Settings to distinguish those variables from the rest of the code.


## Known Issues
(None)

## Developer Notes
* The settings are hard-coded, in the User Settings region of the py file.
* When a span is complete, the span is added to the bake only if the minimum temperature of the span meets the desired temperature.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2021-01-21
### Added
- support `-f` option for fahrenheit in get_temp.sh and read-loop.sh.

### Changed
- Move business logic and related variables including settings to a class.
- Move settings to a dict named `settings` and rename them as follows:
  - `desired_degrees` to `target`.
  - `desired_format` to `scale`.
  - `interval_seconds` to `interval`.
  - `desired_total_seconds` to `minTime`.
  - `desired_comparison` to `compareOp`.
  - `desired_collate_method` to `useStat`.
- Use ccwienk/temper in TemperatureSanitizer.py for improved compatibility.


## [git] - 2021-01-20
### Added
- Support more devices using ccwienk/temper, a fork of urwen/temper, but
  only in get_temp.sh (which read-loop.sh uses if hid-query isn't
  present)
  - Tested with 413d:2107 (applies to [Jepeak Temper High Accurate USB Thermometer Temperature Sensor Data Logger Record for PC Laptop](https://www.amazon.com/gp/product/B009YRP906/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) and others)

### Changed
- Move changes to Changelog.


## [git] - 2017-04-09
### Changed
- Display remaining time while temperature is >= minimum
- Display system time when finished


## [git] - 2017-04-08
### Added
- Add a region called User Settings to distinguish those variables from
  the rest of the code.

### Changed
- Renamed variables for clarity



tempermgr.py
#!/usr/bin/env python

from __future__ import print_function

try:
    input=raw_input
except NameError:
    # Python 3
    pass


install_msg = '''
You do not have ccwienk/temper installed (not to be confused with the HTML templating engine named temper on pypi).
Please run the following commands in command line.
- If you are in Windows first run:
  cd C:\Python*
  or whatever is your Python directory.
- If you are using a Linux-like environment first run:
  sudo su -
  # you don't really need to sudo nor run as root (the program will show a message on how to add a udev rule if there is a PermissionError)
  # you don't really need to use --user if you use a virtualenv, the preferred install method
  # If are using a virtualenv such as creating using the
  # get_temp.sh script, there is no need to run any of these
  # install commands. Instead, run this script using
  #   ~/venv/temper/bin/python
  # OR activate the virtualenv before running via:
  #   source ~/venv/temper/bin/activate

python -m pip install --user --upgrade pip
python -m pip install --user --upgrade pip wheel
python -m pip install --user --upgrade https://github.com/ccwienk/temper/archive/master.zip
'''
try_temperusb = False
_enable_temperusb = False
if try_temperusb:
    try:
        from temperusb import TemperDevice, TemperHandler
    except ImportError:
        print(install_msg)
        _enable_temperusb = False


from temper import Temper

no_dev_msg = '''
No TEMPer device found.
You must use a device supported by temper.
'''


opposite_operator = dict()
opposite_operator["<"] = ">="
opposite_operator["<="] = ">"
opposite_operator[">"] = "<="
opposite_operator[">="] = "<"

def f_to_c(f):
    return (f - 32) / 1.8


def c_to_f(c):
    return (c * 1.8) + 32


import getpass
user = getpass.getuser()
tryGroup = 'driverdev'

def permission_help():
    print("You must run with sudo, root, or add a udev rule via:")
    print()
    print("sudo groupadd {}".format(tryGroup))
    print('echo \'KERNEL=="hidraw*", SUBSYSTEM=="hidraw", MODE="0664",'
          ' GROUP="{}"\' | sudo tee /etc/udev/rules.d/'
          '99-hidraw-permissions.rules'.format(tryGroup))
    print('sudo usermod -a -G {} {}'.format(tryGroup, user))
    print('sudo udevadm control --reload-rules')
    print()
    print('# Then log out and back in if you\'re {}.'.format(user))


class TSBake:
    def __init__(self):
        self.temperatures = list()
        self.time = 0
        # ^ total_seconds
        self.warmTime = None


class TSSpan:
    '''
    An instance of this class holds the cumulative data for one given
    timespan based on the interval setting in seconds.
    '''

    def __init__(self):
        self.count = 0
        # ^ current_span_temperatures_count
        self.total = 0
        # ^ current_span_temperatures_total_temperature
        self.values = []
        # ^ current_temperatures


class TemperDeviceMgr:

    def __init__(self, tempermgr):
        self.tempermgr = tempermgr
        if tempermgr is None:
            raise RuntimeError("The TemperDeviceMgr class requires a"
                               " TemperMgr but the argument was None.")
        if _enable_temperusb:
            self.th = None
            try:
                self.th = TemperHandler()
            except:
                print("A TEMPerV1 compatible device is not plugged in,"
                      " or you have no permission to the usb device."
                      " Try running this script with sudo (sudo python"
                      " TemperatureSanitizer.py) or search online for"
                      " linux usb permissions.")
                exit(2)
            self.tds = self.th.get_devices()
            if len(self.tds) < 1:
                raise RuntimeError(no_dev_msg)
        self.tmpr = Temper()


    def getTemp(self, deviceIndex=0):
        '''
        Get the current temperature of the given device.

        raises:
        PermissionError (see the permission_help function)
        ValueError (bad settings['scale'])
        '''
        if _enable_temperusb:
            return self.tds[deviceIndex].get_temperature(
                format=self.tempermgr.get('scale')
            )
        devices = self.tmpr.read()
        c = devices[0]['internal temperature']
        if self.tempermgr.isF():
            return c_to_f(c)
        elif self.tempermgr.isC():
            return c
        else:
            scale = self.tempermgr.get('scale')
            raise ValueError("The scale {} is unknown.".format(scale))


class TemperMgr:

    def __init__(self, settings=None):
        self.defaults = {}
        self.settings = {}
        self.settings["target"] = 120
        # ^ desired_degrees
        self.settings["scale"] = "fahrenheit"
        # ^ desired_format
        self.settings["interval"] = 60
        # ^ interval_seconds
        self.settings["minTime"] = 120 * 60
        # ^ desired_total_seconds
        self.settings["compareOp"] = ">="
        # ^ desired_comparison
        self.settings['useStat'] = "average"
        # ^ desired_collate_method
        for k,v in self.defaults.items():
            self.settings[k] = v
        if settings is not None:
            for k,v in settings.items():
                self.settings[k] = settings[k]
        self._reinit()

    def _reinit(self):
        self.devicemgr = TemperDeviceMgr(self)
        self.process_term = "Bake"
        if self.settings["compareOp"] in ["<=","<"]:
            self.process_term = "Chill"
            if self.settings['useStat'] in ["max", "maximum"]:
                print("# WARNING: using min or average is recommended\n"
                      "# for chill,\n" +
                      "#   but you selected "
                      + self.settings['useStat'])

        self.complete_bakes = []
        self.incomplete_bakes = []

        self.bake = TSBake()
        # ^ current_bake
        self.warmTime = 0
        # ^ warmup_seconds
        self.span = TSSpan()

    def get(self, name):
        '''
        Get the value of a setting.

        Sequential arguments:
        name -- the name of the setting variable
        '''
        return self.settings.get(name)

    def isF(self):
        scale = self.get('scale')
        if scale.lower() == "fahrenheit":
            return True
        elif scale.lower() == "f":
            return True
        return False

    def isC(self):
        scale = self.get('scale')
        if scale.lower() == "celcius":
            return True
        elif scale.lower() == "c":
            return True
        return False

    def getTemp(self, deviceIndex=0):
        '''
        Get the temperature of the given device.

        raises:
        PermissionError (see the permission_help function)
        ValueError (if settings['scale'] is bad)
        '''
        return self.devicemgr.getTemp(deviceIndex)
        # ^ uses device 0 by default

    def iterate(self, passed, callback=print):
        '''
        Read the temperature and accumulate data into timespans then
        into bake(s).

        This is not a real iterator nor generator.
        Just run it every passed second(s) until it raises
        StopIteration.

        Sequential arguments:
        passed -- This must be the number of seconds that passed since
                  the previous call or no call to iterate. It should
                  usually be the same for every call in the session,
                  even the first call, since it represents the
                  theoretical "slice" of time during which a temperature
                  was present (inaccuracies with that practice are
                  mitigated using an interval setting greater than
                  passed).

        Raises:
        PermissionError (See the permission_help function)
        ValueError (if settings['scale'] is bad)
        StopIteration (if the bake is complete)
        '''
        # try:
        this_temp = self.getTemp()
        '''
        except:
            callback("#Couldn't finish reading temperature." +
                     "#  The device seems to be disconnected.")
            callback("exit_datetime: "+str(datetime.datetime.now()))
            callback("exit_bake_seconds: "+str(self.bake.time))
            callback("#exit_bake_minutes: "+str(self.bake.time/60))
            raise RuntimeError("The device is not accessible.")
        '''
        self.span.count += 1
        self.span.total += this_temp
        self.span.values.append(this_temp)
        if self.span.count >= self.get("interval"):
            this_avg = self.span.total / self.span.count
            this_min = min(self.span.values)
            good_value = self.is_criteria_met(self.span.values)
            del self.span.values[:]
            met_msg = ("("
                       + opposite_operator[self.settings["compareOp"]]
                       + " " + str(self.settings["target"]) + ") ")
            # if (this_min >= self.settings["target"]):
            if good_value is not None:
                self.bake.temperatures.append(good_value)
                self.bake.time += self.span.count
                met_msg = ("(" + self.settings["compareOp"] + " "
                           + str(self.settings["target"]) + ") ")
                if (self.bake.time >= self.settings["minTime"]):
                    self.bake.warmTime = self.warmTime
                    self.complete_bakes.append(self.bake)
                    self.bake = TSBake()
            else:
                if len(self.bake.temperatures) > 0:
                    if (self.bake.time < self.settings["minTime"]):
                        self.bake.warmTime = self.warmTime
                        self.incomplete_bakes.append(self.bake)
                        self.bake = TSBake()
                    else:
                        message = (
                            "#Logic error detected (this should never"
                            " happen): program did not end when "
                            + self.process_term.lower()
                            + " was successful (appending "
                            + self.process_term.lower()
                            + " to complete_"
                            + self.process_term.lower() + "s anyway)."
                        )
                        self.bake.warmTime = self.warmTime
                        print("#current_pre{}_minutes: {}"
                              "".format(self.process_term.lower(),
                                        self.bake.warmTime/60))
                        self.complete_bakes.append(self.bake)
                        self.bake = TSBake()
            sustained_msg = ""
            if (self.bake is not None) and (self.bake.time > 0):
                # sustained_msg = (
                # "  # sustained:{}m".format(self.bake.time/60)
                # )
                sustained_msg = "  # remaining:{}m".format(
                    (self.settings["minTime"]-self.bake.time)/60
                )
            else:
                sustained_msg = ("  # pre{}_mins:"
                                 "".format(self.process_term.lower(),
                                           self.warmTime/60))
            print("#" + met_msg + "Last "
                  + str(self.span.count)
                  + " second(s) avg:" + str(this_avg) + "; min:"
                  + str(this_min) + sustained_msg)
            self.span.count = 0
            self.span.total = 0
            if len(self.complete_bakes) > 0:
                print("#"+self.process_term+" is finished: ")
                raise StopIteration
        self.warmTime += passed

    def is_criteria_met(self, temperatures):
        # formerly is_criteria_met
        result = None
        operand = None
        if self.settings['useStat'] in ["min","minimum"]:
            operand = min(temperatures)
        elif self.settings['useStat'] in ["max","maximum"]:
            operand = max(temperatures)
        elif self.settings['useStat'] in ["avg","average"]:
            operand = sum(temperatures) / float(len(temperatures))
        else:
            print("# settings['useStat'] must be: \n" +
                  "#   min, max, or average")
            exit(5)
        met_enable = False
        if self.settings["compareOp"] == ">=":
            met_enable = (operand >= self.settings["target"])
        elif self.settings["compareOp"] == ">":
            met_enable = (operand > self.settings["target"])
        elif self.settings["compareOp"] == "<=":
            met_enable = (operand <= self.settings["target"])
        elif self.settings["compareOp"] == "<":
            met_enable = (operand < self.settings["target"])
        else:
            print("# Unknown comparison was selected. Please use: \n" +
                  "#   <, <=, >, or >=")
            exit(6)
        if met_enable:
            result = operand
        return result
